src/pcp.rs

```diff
@@ -0,0 +1,596 @@
+use crate::polynomial::{poly_deg, poly_eval, poly_mul};
+
+pub mod gadgets;
+pub mod types;
+
+/// Errors propagagted by methods in this module.
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum PcpError {
+    /// The caller of an arithmetic circuit provided the wrong number of inputs. This error may
+    /// occur when evaluating a validity circuit or gadget.
+    #[error("wrong number of inputs to arithmetic circuit")]
+    CircuitInLen,
+
+    /// This error is returned by `collect` if the input slice is empty.
+    #[error("collect requires at least one input")]
+    CollectInLen,
+
+    /// This error is returned by `collect` if the two or more verifier shares have different
+    /// gadget arities.
+    #[error("collect inputs have mismatched gadget arity")]
+    CollectGadgetInLenMismatch,
+
+    /// Returned if an FFT operation propagates an error.
+    #[error("FFT error")]
+    Fft(#[from] FftError),
+
+    /// When evaluating a gadget on polynomials, this error is returned if the input polynomials
+    /// don't all have the same length.
+    #[error("gadget called on polynomials with different lengths")]
+    GadgetPolyInLen,
+
+    /// When evaluating a gadget on polynomials, this error is returned if the slice allocated for
+    /// the output polynomial is too small.
+    #[error("sliice allocated for gadget output is too small")]
+    GadgetPolyOutLen,
+
+    /// The proof string is either too large or too short for the given type.
+    #[error("attempted query on proof with invalid length")]
+    QueryProofLen,
+}
+
+/// A data type. Implementations of this trait specify an arithmetic circuit that determines
+/// whether a given value is valid.
+pub trait Type<F, G>: Sized
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    /// Evalauates the validity circuit on the given input (i.e., `self`) and returns the output.
+    /// Slice `rand` is the random input consumed by the validity circuit.
+    ///
+    /// ```
+    /// use prio::pcp::types::Boolean;
+    /// use prio::pcp::Type;
+    /// use prio::field::{FieldElement, Field64};
+    ///
+    /// type F = Field64;
+    /// type T = Boolean<F>;
+    ///
+    /// let x = T::new(false);
+    ///
+    /// let mut rand: Vec<F> = Vec::with_capacity(x.valid_rand_len());
+    /// for _ in 0..rand.len() {
+    ///     rand.push(F::rand());
+    /// }
+    ///
+    /// let v = x.valid(&mut x.gadget(0), &rand).unwrap();
+    /// assert_eq!(v, F::zero());
+    /// ```
+    fn valid(&self, g: &mut dyn GadgetCallOnly<F>, rand: &[F]) -> Result<F, PcpError>;
+
+    /// Returns a reference to the underlying data.
+    fn as_slice(&self) -> &[F];
+
+    /// Constructs an instance of this type from the given data without checking whether the data
+    /// is valid. This method takes ownership of `data`.
+    fn clone_with(&self, data: Vec<F>) -> Self;
+
+    /// The length of the random input expected by the validity circuit.
+    fn valid_rand_len(&self) -> usize;
```

Comment:
All of these calls might, in theory, depend on the input length. This is why we pass an instance of the type to each of these functions. But I agree with what you're getting at here. Our goal is to rework the code to express the fact that `valid_rand_len`, `valid_gadget_calls`, and `gadget` are all parameters of the type and not the underlying data.

As we discussed on Friday afternoon, we don't yet have an example of a validity circuit that would take more than 1 random field element, so I'm comfortable making changing `valid_rand_len(&self) -> size` to `valid_rand_len() -> size`. However, it's going to be common for the output of `valid_gadget_calls` and `gadget` to depend on the input length. (E.g., see `PolyCheckedVector` in #30.)

So the question is how to express "length". We can't use const generics because the length might not be known until runtime. One alternative is to add a `len(&self) -> usize` method that returns the "length" of the underlying input. We would then pass this length as an input to `valid_rand_len`, `valid_gadget_calls`, and `gadget`. This is a bit ugly, since not every type has a natural notion of "length". But since it's the type that defines what "length" means, this isn't really that big of a deal.

Reply:
It seems like anywhere that a caller might use `valid_rand_len`, they will probably also be using either or both of `gadget` and `valid_gadget_calls`, and if those require an instance of `Value`, then `valid_rand_len` taking an instance isn't any more of a burden for the caller. So let's leave `valid_rand_len` as is.